File: shanoir-ng-studies/src/main/java/org/shanoir/ng/subject/dto/SubjectDTO.java
/**
 * Shanoir NG - Import, manage and share neuroimaging data
 * Copyright (C) 2009-2019 Inria - https://www.inria.fr/
 * Contact us on https://project.inria.fr/shanoir/
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/gpl-3.0.html
 */

package  org.shanoir.ng.subject.dto;

import java.time.LocalDate;
import java.util.List;

import org.shanoir.ng.shared.dateTime.LocalDateAnnotations;
import org.shanoir.ng.subject.model.HemisphericDominance;
import org.shanoir.ng.subject.model.ImagedObjectCategory;
import org.shanoir.ng.subject.model.Sex;
import org.shanoir.ng.subjectstudy.dto.SubjectStudyDTO;


public class SubjectDTO {
 
    private Long id;
    
    private String name;
    
    private String identifier;

    @LocalDateAnnotations
    private LocalDate birthDate;

    private HemisphericDominance languageHemisphericDominance;

    private HemisphericDominance manualHemisphericDominance;

    private ImagedObjectCategory imagedObjectCategory;

    private Sex sex;
    
    private List<SubjectStudyDTO> subjectStudyList;
    
    private boolean preclinical;

	/**
	 * @return the id
	 */
	public Long getId() {
		return id;
	}

	/**
	 * @param id the id to set
	 */
	public void setId(Long id) {
		this.id = id;
	}

	/**
	 * @return the name
	 */
	public String getName() {
		return name;
	}

	/**
	 * @param name the name to set
	 */
	public void setName(String name) {
		this.name = name;
	}

	/**
	 * @return the identifier
	 */
	public String getIdentifier() {
		return identifier;
	}

	/**
	 * @param identifier the identifier to set
	 */
	public void setIdentifier(String identifier) {
		this.identifier = identifier;
	}

	/**
	 * @return the birthDate
	 */
	public LocalDate getBirthDate() {
		return birthDate;
	}

	/**
	 * @param birthDate the birthDate to set
	 */
	public void setBirthDate(LocalDate birthDate) {
		this.birthDate = birthDate;
	}

	/**
	 * @return the languageHemisphericDominance
	 */
	public HemisphericDominance getLanguageHemisphericDominance() {
		return languageHemisphericDominance;
	}

	/**
	 * @param languageHemisphericDominance the languageHemisphericDominance to set
	 */
	public void setLanguageHemisphericDominance(HemisphericDominance languageHemisphericDominance) {
		this.languageHemisphericDominance = languageHemisphericDominance;
	}

	/**
	 * @return the manualHemisphericDominance
	 */
	public HemisphericDominance getManualHemisphericDominance() {
		return manualHemisphericDominance;
	}

	/**
	 * @param manualHemisphericDominance the manualHemisphericDominance to set
	 */
	public void setManualHemisphericDominance(HemisphericDominance manualHemisphericDominance) {
		this.manualHemisphericDominance = manualHemisphericDominance;
	}

	/**
	 * @return the imagedObjectCategory
	 */
	public ImagedObjectCategory getImagedObjectCategory() {
		return imagedObjectCategory;
	}

	/**
	 * @param imagedObjectCategory the imagedObjectCategory to set
	 */
	public void setImagedObjectCategory(ImagedObjectCategory imagedObjectCategory) {
		this.imagedObjectCategory = imagedObjectCategory;
	}

	/**
	 * @return the sex
	 */
	public Sex getSex() {
		return sex;
	}

	/**
	 * @param sex the sex to set
	 */
	public void setSex(Sex sex) {
		this.sex = sex;
	}

	/**
	 * @return the subjectStudyList
	 */
	public List<SubjectStudyDTO> getSubjectStudyList() {
		return subjectStudyList;
	}

	/**
	 * @param subjectStudyList the subjectStudyList to set
	 */
	public void setSubjectStudyList(List<SubjectStudyDTO> subjectStudyList) {
		this.subjectStudyList = subjectStudyList;
	}
	
	public boolean isPreclinical() {
		return preclinical;
	}

	public void setPreclinical(boolean preclinical) {
		this.preclinical = preclinical;
	}
    
//  private List<UserPersonalCommentSubject> userPersonalCommentList = new ArrayList<UserPersonalCommentSubject>(0);

//    private PseudonymusHashValues pseudonymusHashValues;

//    private List<Dataset> datasetList = new ArrayList<Dataset>(0);
//
//    private List<Examination> examinationList = new ArrayList<Examination>(0);
//
//    private List<RelSubjectGroupOfSubjects> relSubjectGroupOfSubjectsList = new ArrayList<RelSubjectGroupOfSubjects>(0);
//
//    private List<UserPersonalCommentSubject> userPersonalCommentList = new ArrayList<UserPersonalCommentSubject>(0);

    
    
}

File: shanoir-ng-front/src/app/subjects/shared/subject.dto.ts
/**
 * Shanoir NG - Import, manage and share neuroimaging data
 * Copyright (C) 2009-2019 Inria - https://www.inria.fr/
 * Contact us on https://project.inria.fr/shanoir/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/gpl-3.0.html
 */
import { Injectable } from '@angular/core';

import { Examination } from '../../examinations/shared/examination.model';
import { Id } from '../../shared/models/id.model';
import { StudyDTOService } from '../../studies/shared/study.dto';
import { Tag } from '../../tags/tag.model';
import { ImagedObjectCategory } from './imaged-object-category.enum';
import { SubjectStudyDTO } from './subject-study.dto';
import { Subject } from './subject.model';
import {Sex, SubjectType} from './subject.types';
import {formatDate} from "@angular/common";
import {QualityTag} from "../../study-cards/shared/quality-card.model";
import {SimpleStudy, Study} from "../../studies/shared/study.model";


@Injectable()
export class SubjectDTOService {

    constructor(
    ) {}

    /**
     * Convert from DTO to Entity
     * Warning : DO NOT USE THIS IN A LOOP, use toEntityList instead
     * @param result can be used to get an immediate temporary result without waiting async data
     */
    public toEntity(dto: SubjectDTO, result?: Subject): Promise<Subject> {
        if (!result) result = new Subject();
        SubjectDTOService.mapSyncFields(dto, result);
        return Promise.resolve(result);
    }

    /**
     * Convert from a DTO list to an Entity list
     * @param result can be used to get an immediate temporary result without waiting async data
     */
    public toEntityList(dtos: SubjectDTO[], result?: Subject[]): Promise<Subject[]>{
        if (!result) result = [];
        if (dtos) {
            for (let dto of dtos) {
                let entity = new Subject();
                SubjectDTOService.mapSyncFields(dto, entity);
                result.push(entity);
            }
        }
        return Promise.resolve(result);
    }

    static mapSyncFields(dto: SubjectDTO, entity: Subject): Subject {
        entity.id = dto.id;
        entity.examinations = dto.examinations ? dto.examinations.map(examId => {
            let exam: Examination = new Examination();
            exam.id = examId.id;
            return exam;
        }) : null;
        entity.name = dto.name;
        entity.identifier = dto.identifier;
        entity.birthDate = dto.birthDate ? new Date(dto.birthDate) : null;
        entity.preclinical = dto.preclinical;
        entity.languageHemisphericDominance = dto.languageHemisphericDominance;
        entity.manualHemisphericDominance = dto.manualHemisphericDominance;
        entity.imagedObjectCategory = dto.imagedObjectCategory;
        entity.sex = dto.sex;
        entity.studyIdentifier = dto.studyIdentifier;
        entity.isAlreadyAnonymized = dto.isAlreadyAnonymized;
        entity.subjectType = dto.subjectType;
        entity.physicallyInvolved = dto.physicallyInvolved;
        entity.tags = dto.tags;
        entity.qualityTag = dto.qualityTag;
        entity.study = new Study()
        entity.study.id = dto.studyId;
        return entity;
    }

    static tagDTOToTag(tagDTO: any): Tag {
        let tag: Tag = new Tag();
        tag.id = tagDTO.id;
        tag.name = tagDTO.name;
        tag.color = tagDTO.color;
        return tag;
    }
}

export class SubjectDTO {

    id: number;
    examinations: Id[];
    name: string;
    identifier: string;
    birthDate: string;
    languageHemisphericDominance: "Left" | "Right";
    manualHemisphericDominance: "Left" | "Right";
    imagedObjectCategory: ImagedObjectCategory;
    sex: Sex;
    selected: boolean = false;
    preclinical: boolean;
    studyIdentifier: string;
    isAlreadyAnonymized: boolean = false;
    subjectType: SubjectType;
    physicallyInvolved: boolean;
    tags: Tag[];
    qualityTag: QualityTag;
    study: SimpleStudy;
    studyId: number;

    constructor(subject: Subject) {
        this.id = subject.id;
        if (subject.examinations) this.examinations = Id.toIdList(subject.examinations);
        this.name = subject.name;
        this.identifier = subject.identifier;
        if (subject.birthDate && !isNaN(subject.birthDate.getTime())) this.birthDate = formatDate(subject.birthDate, 'yyyy-MM-dd', 'en');
        this.languageHemisphericDominance = subject.languageHemisphericDominance;
        this.manualHemisphericDominance = subject.manualHemisphericDominance;
        this.imagedObjectCategory = subject.imagedObjectCategory;
        this.sex = subject.sex;
        this.selected = subject.selected;
        this.preclinical = subject.preclinical;
        this.studyIdentifier = subject.studyIdentifier;
        this.isAlreadyAnonymized = subject.isAlreadyAnonymized;
        this.subjectType = subject.subjectType;
        this.physicallyInvolved = subject.physicallyInvolved;
        this.tags = subject.tags;
        this.qualityTag = subject.qualityTag;
        this.study = new SimpleStudy(subject.study);
    }
}
